service/history/shard/context_impl.go

```diff
@@ -1590,81 +1576,88 @@ func (s *ContextImpl) transitionLocked(request contextRequest) {
 
 	setStateStopped := func() {
 		s.state = contextStateStopped
+		// Do this again in case we skipped the stopping state, which could happen
+		// when calling CloseShardByID or the controller is shutting down.
+		s.lifecycleCancel()
 	}
 
 	switch s.state {
 	case contextStateInitialized:
 		switch request.(type) {
 		case contextRequestAcquire:
 			setStateAcquiring()
-			return
+			return nil
 		case contextRequestStop:
 			setStateStopping()
-			return
+			return nil
 		case contextRequestFinishStop:
 			setStateStopped()
-			return
+			return nil
 		}
 	case contextStateAcquiring:
-		switch request.(type) {
+		switch request := request.(type) {
 		case contextRequestAcquire:
-			return // nothing to do, already acquiring
+			return nil // nothing to do, already acquiring
 		case contextRequestAcquired:
 			s.state = contextStateAcquired
-			return
+			if request.engine != nil {
+				// engineFuture.Set should only be called inside stateLock when state is
+				// Acquiring, so that other code (i.e. finishStop) can know that after a state
+				// transition to Stopping/Stopped, engineFuture cannot be Set.
+				if s.engineFuture.Ready() {
+					// defensive check, this should never happen
+					return errInvalidTransition
+				}
+				s.engineFuture.Set(request.engine, nil)
+			}
+			return nil
 		case contextRequestLost:
-			return // nothing to do, already acquiring
+			return nil // nothing to do, already acquiring
 		case contextRequestStop:
 			setStateStopping()
-			return
+			return nil
 		case contextRequestFinishStop:
 			setStateStopped()
-			return
+			return nil
 		}
 	case contextStateAcquired:
 		switch request.(type) {
 		case contextRequestAcquire:
-			return // nothing to to do, already acquired
+			return nil // nothing to to do, already acquired
 		case contextRequestLost:
 			setStateAcquiring()
-			return
+			return nil
 		case contextRequestStop:
 			setStateStopping()
-			return
+			return nil
 		case contextRequestFinishStop:
 			setStateStopped()
-			return
+			return nil
 		}
 	case contextStateStopping:
 		switch request.(type) {
 		case contextRequestStop:
 			// nothing to do, already stopping
-			return
+			return nil
 		case contextRequestFinishStop:
 			setStateStopped()
-			return
+			return nil
 		}
 	}
 	s.contextTaggedLogger.Warn("invalid state transition request",
 		tag.ShardContextState(int(s.state)),
 		tag.ShardContextStateRequest(fmt.Sprintf("%T", request)),
 	)
+	return errInvalidTransition
```

Comment:
maybe give this error tags for current and requested states to aid in debugging?

Reply:
The log message has tags. I don't think adding them to the error value adds anything